htlcswitch/held_htlc_set.go

```diff
@@ -0,0 +1,59 @@
+package htlcswitch
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+)
+
+type heldHtlcSet struct {
+	// set keeps track of outstanding intercepted forwards.
+	set map[channeldb.CircuitKey]InterceptedForward
+}
+
+func newHeldHtlcSet() *heldHtlcSet {
+	return &heldHtlcSet{
+		set: make(map[channeldb.CircuitKey]InterceptedForward),
+	}
+}
+
+func (h *heldHtlcSet) forEach(cb func(InterceptedForward)) {
+	for _, fwd := range h.set {
+		cb(fwd)
+	}
+}
+
+func (h *heldHtlcSet) popAll(cb func(InterceptedForward)) {
+	for _, fwd := range h.set {
+		cb(fwd)
+	}
+
+	h.set = make(map[channeldb.CircuitKey]InterceptedForward)
+}
+
+func (h *heldHtlcSet) pop(key channeldb.CircuitKey) (InterceptedForward, error) {
+	intercepted, ok := h.set[key]
+	if !ok {
+		return nil, fmt.Errorf("fwd %v not found", key)
+	}
+
+	delete(h.set, key)
+
+	return intercepted, nil
+}
+
+func (h *heldHtlcSet) exists(key channeldb.CircuitKey) bool {
+	_, ok := h.set[key]
+
+	return ok
+}
+
+func (h *heldHtlcSet) push(key channeldb.CircuitKey,
+	fwd InterceptedForward) {
+
+	if h.exists(key) {
+		panic("htlc already exists in set")
```

Comment:
Why panic over returning an error?

Reply:
This seemed to me an error in the category "index out of bounds", which also panics with the default go types. But can make it an error.

Reply:
I am not sure what the best way is. For a slice out of bounds, there is also no pre-check and `log.Critical`.

@Crypt-iQ what do you prefer here?

Reply:
It's nicer to do a graceful shutdown, though robust code should be able to tolerate a panic. If the code isn't robust in other areas (e.g. https://github.com/lightningnetwork/lnd/issues/6704) and lnd panics, maybe you get some impossible database states/corruption which is both good and bad. Good because you can notice and fix them. Bad for the user. So I don't really have a strong opinion.

Reply:
Added error return value. Maybe it is not primitive enough to be comparable with a slice index out of bounds.